agera/src/main/java/com/google/android/agera/BaseObservable.java

```diff
@@ -38,149 +39,131 @@
+        throw new IllegalStateException("Updatable already added, cannot add.");
       }
-    }
-
-    void dispatchUpdate() {
-      if (!handler.hasMessages(MSG_UPDATE, this)) {
-        handler.obtainMessage(MSG_UPDATE, this).sendToTarget();
+      if (updatablesAndHandlers[index] == null) {
+        indexToAdd = index;
       }
     }
-
-    private void add(@NonNull final Updatable updatable, @NonNull final Handler handler) {
-      boolean added = false;
-      for (int index = 0; index < updatablesAndHandlers.length; index += 2) {
-        if (updatablesAndHandlers[index] == updatable) {
-          throw new IllegalStateException("Updatable already added, cannot add.");
-        }
-        if (updatablesAndHandlers[index] == null && !added) {
-          updatablesAndHandlers[index] = updatable;
-          updatablesAndHandlers[index + 1] = handler;
-          added = true;
-        }
-      }
-      if (!added) {
-        final int newIndex = updatablesAndHandlers.length;
-        updatablesAndHandlers = Arrays.copyOf(updatablesAndHandlers,
-            Math.max(newIndex * 2, newIndex + 2));
-        updatablesAndHandlers[newIndex] = updatable;
-        updatablesAndHandlers[newIndex + 1] = handler;
-      }
-      size++;
+    if (indexToAdd == -1) {
+      indexToAdd = updatablesAndHandlers.length;
+      updatablesAndHandlers = Arrays.copyOf(updatablesAndHandlers,
+          indexToAdd < 2 ? 2 : indexToAdd * 2);
     }
+    updatablesAndHandlers[indexToAdd] = updatable;
+    updatablesAndHandlers[indexToAdd + 1] = handler;
+    size++;
+  }
 
-    private void remove(@NonNull final Updatable updatable) {
-      for (int index = 0; index < updatablesAndHandlers.length; index += 2) {
-        if (updatablesAndHandlers[index] == updatable) {
-          ((WorkerHandler) updatablesAndHandlers[index + 1]).removeMessages(
-              WorkerHandler.MSG_CALL_UPDATABLE, updatable);
-          updatablesAndHandlers[index] = null;
-          updatablesAndHandlers[index + 1] = null;
-          size--;
-          return;
-        }
+  private void remove(@NonNull final Updatable updatable) {
+    for (int index = 0; index < updatablesAndHandlers.length; index += 2) {
+      if (updatablesAndHandlers[index] == updatable) {
+        ((WorkerHandler) updatablesAndHandlers[index + 1]).removeMessages(
+            WorkerHandler.MSG_CALL_UPDATABLE, updatable);
+        updatablesAndHandlers[index] = null;
+        updatablesAndHandlers[index + 1] = null;
+        size--;
+        return;
       }
-      throw new IllegalStateException("Updatable not added, cannot remove.");
     }
+    throw new IllegalStateException("Updatable not added, cannot remove.");
+  }
 
-    synchronized void sendUpdate() {
+  synchronized void sendUpdate() {
+    handler.removeMessages(WorkerHandler.MSG_UPDATE, this);
+    final long elapsedRealtimeMillis =
+        shortestUpdateWindowMillis > 0 ? elapsedRealtime() : 0;
```

Comment:
Despite the logical correctness involving `shortestUpdateWindowMillis <= 0`, I would prefer elapsedRealtimeMillis and lastUpdateTimestamp updated as usual, so I don't run into difficulty reading the following line:

```
timeFromLastUpdate = elapsedRealtimeMillis - lastUpdateTimestamp
                   = 0 - some timestamp
                   = negative result
```

Or, if your concern is about calling elapsedRealtime() unnecessarily, then a short-cutting if-block is better:

```
if (shortestUpdateWindowMillis > 0) {
  final long elapsedRealtimeMillis = elapsedRealtime();
  final long timeFromLastUpdate = elapsedRealtimeMillis - lastUpdateTimestamp;
  if (timeFromLastUpdate  < shortestUpdateWindowMillis) {
    handler.sendMessageDelayed(handler.obtainMessage(WorkerHandler.MSG_UPDATE, this),
        shortestUpdateWindowMillis - timeFromLastUpdate);
    return;
  }
  lastUpdateTimestamp = elapsedRealtimeMillis;
}
// dispatch updates
for (int index = 0; ...
```